File: code/python/parse/simple_parser.py
'''
Created on Jun 19, 2020

@author: tvandrun
'''
from tokenizer import Tokenizer

class BadSyntax(Exception):
    def __init__(self, message):
        self.message = message

class SyntaxTree :
    def __init__(self, nonterminal, children):
        self.nonterminal = nonterminal
        self.children = children
    def __str__(self):
        return ("(" + self.nonterminal + " [" + 
                ",".join([str(x) for x in self.children]) + "]" )

def check_token(tok, ty):
    if tok.pos != ty : raise BadSyntax("Expected " + ty + ", found " + tok.pos)

def get_token(tokzer, ty):
    token = tokzer.extract_current()
    check_token(token, ty)
    return token
    
def sentence(tokzer):
    return SyntaxTree("Sentence", [noun_phrase(tokzer), predicate(tokzer)])
    
def noun_phrase(tokzer):
    return SyntaxTree("NounPhrase", 
                      [get_token(tokzer, "ART"), adjective_list(tokzer), 
                       get_token(tokzer, "NOUN")])
    
def adjective_list(tokzer):
    adj_list = []
    while tokzer.current().pos == "ADJ" :
        adj_list.append(tokzer.extract_current())
    return SyntaxTree("AdjList", adj_list)

def predicate(tokzer):
    return SyntaxTree("Predicate", [verb_phrase(tokzer), verb_modifier(tokzer)])

def verb_phrase(tokzer):    
    first_pos = tokzer.current().pos
    if first_pos == "LV" :
        return linking_verb_phrase(tokzer)
    elif first_pos == "TV" :
        return transitive_verb_phrase(tokzer)
    elif first_pos == "IV" :
        return intransitive_verb_phrase(tokzer)
    else :
        raise BadSyntax("Expected LV, TV, or IV, found " + first_pos)
    
def linking_verb_phrase(tokzer):
    return SyntaxTree("LinkingVerbPhrase", 
                      [get_token(tokzer, "LV"), get_token(tokzer, "ADJ")])

def transitive_verb_phrase(tokzer):
    return SyntaxTree("TransitiveVerbPhrase",
                      [get_token(tokzer, "TV"), noun_phrase(tokzer)])

def intransitive_verb_phrase(tokzer):
    return SyntaxTree("IntransitiveVerbPhrase",
                      [get_token(tokzer, "IV")])

def verb_modifier(tokzer):
    if tokzer.has_current() :
        first_pos = tokzer.current().pos
        if first_pos == "PREP" :
            return prep_phrase(tokzer)
        elif first_pos == "ADV" :
            return tokzer.extract_current()
    return None

def prep_phrase(tokzer):
    return SyntaxTree("PrepPhrase", [get_token(tokzer, "PREP"), noun_phrase(tokzer)])

trivial = "the dog ran"
a = "the big dog ran through the bright field"
b = "the loud dog chased the small ball quickly"

try:
    print(sentence(Tokenizer(b))) 
except Exception as ex :
    print(str(ex.message)) 

File: code/java/examples/rdparse/grammar/SimpleParser.java
package grammar;

import java.util.ArrayList;

import grammar.Token.TokenType;

public class SimpleParser {

    private Tokenizer tokey;
    
    public String parse(String message) {
        tokey = new Tokenizer(message);
        return (new Sentence()).toString();
    }
    
    private static String trivial = "the dog ran";
    private static String a = "the big dog ran through the bright field";
    private static String b = "the loud dog chased the small ball quickly";

    
    public static void main(String[] args) {
        Parser parsy = new Parser();
        System.out.println(parsy.parse(a));
    }
   
    class Sentence {
        NounPhrase subject;
        Predicate pred;
        Sentence() {
            subject = new NounPhrase();
            pred = new Predicate();
        }
        public String toString() { return "(Sentence [" + subject + "," + pred + "])"; }
    }

    class NounPhrase {
        Token art;
        ArrayList<Token> adjs;
        Token nominal;

        public NounPhrase() {
            art = tokey.extractCurrent(TokenType.ART);
            adjs = new ArrayList<Token>();
            while (tokey.current().ty == TokenType.ADJ)
                adjs.add(tokey.extractCurrent(TokenType.ADJ));
            nominal = tokey.extractCurrent(TokenType.NOUN);
        }
        public String toString() { return "(ConcNounPhrase [" + art + ", " +
                adjs + ", " + nominal + "])"; }
    }
    
    
    class Predicate {
        VerbPhrase vp;
        VerbModifier vm;
        public Predicate() {
            switch(tokey.current().ty) {
            case LV: vp = new LinkingVerbPhrase(); break;
            case TV: vp = new TransitiveVerbPhrase(); break;
            case IV: vp = new IntransitiveVerbPhrase(); break;
            default: throw new SyntaxException();
            }
            if (tokey.hasCurrent())
                switch(tokey.current().ty) {
                case PREP: vm = new PrepositionalPhrase (); break;
                case ADV: vm = new Adverbial(); break;
                default: vm = null;
                }
            else vm = null;
        }
        public String toString() { return "(Predicate [" + vp + ", " +
                (vm == null? "None" : vm.toString()) + "])"; 
        }
    }
    
    static interface VerbModifier {}
    
    class PrepositionalPhrase implements VerbModifier {
        Token prep;
        NounPhrase object;
        public PrepositionalPhrase() {
            prep = tokey.extractCurrent(TokenType.PREP);
            object = new NounPhrase();
        }
        public String toString() { return "(PrepPhrase [" + prep + "," + object + "])"; }
    }
    
    class Adverbial implements VerbModifier {

        Token adv;
        public Adverbial() {
            adv = tokey.extractCurrent(TokenType.ADV);
        }
        public String toString() { return adv.toString(); }
    }
    
    static interface VerbPhrase {}
    
    class LinkingVerbPhrase implements VerbPhrase {
        Token lv;
        Token adj;
        public LinkingVerbPhrase() {
            lv = tokey.extractCurrent(TokenType.LV);
            adj = tokey.extractCurrent(TokenType.ADJ);
        }
        public String toString() { return "(LVP [" + lv + ", " + adj + "])"; }
    }
    class TransitiveVerbPhrase implements VerbPhrase {
        Token tv;
        NounPhrase directObject;
        public TransitiveVerbPhrase() {
            tv = tokey.extractCurrent(TokenType.TV);
            directObject = new NounPhrase();
        }
        public String toString() { return "(TVP [" + tv + ", " + directObject + "])"; }
    }
    class IntransitiveVerbPhrase implements VerbPhrase {

        Token iv;
        public IntransitiveVerbPhrase() {
            iv = tokey.extractCurrent(TokenType.IV);
        }
        public String toString() { return "(IVP [" + iv + "])"; }
   }

}
